[Documentation Code Camp] Abstract classes documentation. (#906)
mxcube/mxcubecore: docs/source/dev/abstract_classes.md

# Abstract Classes
## General Concept

> The role of the mxcubecore Abstract Classes is to provide an API to be used by both mxcubeweb and mxcubeqt clients. It defines methods, properties and signals. Most of the methods are defined only to give the footprint and shold be, wneh appropriate, overloaded by the inheriting classes.

Reviewer: Typo: 'when'

Reviewer 2: Typo: 'should'

Author: Wow, second hawk eye. Thanks @AnnieHeroux

> The role of the mxcubecore Abstract Classes is to provide an API to be used by both mxcubeweb and mxcubeqt clients. It defines methods, properties and signals. Most of the methods are defined only to give the footprint and shold be, wneh appropriate, overloaded by the inheriting classes.

Reviewer 2: ```suggestion
The role of the mxcubecore Abstract Classes is to provide an API to be used by both mxcubeweb and mxcubeqt clients. It defines methods, properties and signals. Most of the methods are defined only to give the footprint and should be, when appropriate, overloaded by the inheriting classes.
```

Reviewer 3: @AnnieHeroux Ah! I see now you figured out how to do specific suggestions directly in the code. Great!

Abstract class can be inherited by any class, including another abstract class.

> Any abstract calss, if not inheriting from another abstract class should inherit from the **HardwareObject** class.

Reviewer: Typo: 'class;

> Any abstract calss, if not inheriting from another abstract class should inherit from the **HardwareObject** class.

Reviewer: OR the HardwareObjectYaml class

> Any abstract calss, if not inheriting from another abstract class should inherit from the **HardwareObject** class.

Reviewer 2: ```suggestion
Any abstract class, if not inheriting from another abstract class should inherit from the **HardwareObject** class.
```

> If get_value/set_value methods are defined in the abstract class, the name of the class defines what will be represented. For example get_value/set_value in the AbstractEnergy will read/set the enery, while the wavelength, handled in the same class, will have specific methoods for its reading/setting.

Reviewer 2: typo: enery => energy  methood=> method

I am still learning how to exercise my github muscle. THere might be a better way...

Reviewer 3: > typo: enery => energy methood=> method
> 
> I am still learning how to exercise my github muscle. THere might be a better way...

@AnnieHeroux, I suggest trying the points 5 and 7 of "Starting a review": <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/reviewing-proposed-changes-in-a-pull-request#starting-a-review>

> If get_value/set_value methods are defined in the abstract class, the name of the class defines what will be represented. For example get_value/set_value in the AbstractEnergy will read/set the enery, while the wavelength, handled in the same class, will have specific methoods for its reading/setting.

Reviewer 2: ```suggestion
If get_value/set_value methods are defined in the abstract class, the name of the class defines what will be represented. For example get_value/set_value in the AbstractEnergy will read/set the energy, while the wavelength, handled in the same class, will have specific methods for its reading/setting.
```

## Methods

> There are two methods, which are avaiable for any class, as inherited from HardwareObject - **get_object_by_role()** and **get_property()**. Both methods allow to retrieve information from the configuration file (xml or yaml)

Reviewer: This will hopefully need rewrting soon, but meanwhile, well, OK,

Reviewer 2: Typo: available

> There are two methods, which are avaiable for any class, as inherited from HardwareObject - **get_object_by_role()** and **get_property()**. Both methods allow to retrieve information from the configuration file (xml or yaml)

Reviewer 2: ```suggestion
There are two methods, which are available for any class, as inherited from HardwareObject - **get_object_by_role()** and **get_property()**. Both methods allow to retrieve information from the configuration file (xml or yaml)
```

- get_object_by_role returns an object, associated to a role

  ```<object href="/aperture" role="beamsize"/>```
- get_property returns a property - numeric, string or None

  ```<values>{"IN": False, "OUT": True}</values>```

Methods, specific to different abstract classes will be listed further.

## STATES and SPECIFIC_STATES Enum
There is only a limited number of general states, shared by all HardwareObjects.
They are defined as an Enum as follows and should not be overridden.
```
@enum.unique
class HardwareObjectState(enum.Enum):
    """Enumeration of common states, shared between all HardwareObjects"""

    UNKNOWN = 0
    WARNING = 1
    BUSY = 2
    READY = 3
    FAULT = 4
    OFF = 5
```

> The HardwareObjectState class is asigned to the STATES object, so in the inheriting from HardwareObject classes it is accessible as self.STATES

Reviewer 2: Typo: assigned

> The HardwareObjectState class is asigned to the STATES object, so in the inheriting from HardwareObject classes it is accessible as self.STATES

Reviewer 2: ```suggestion
The HardwareObjectState class is assigned to the STATES object, so in the inheriting from HardwareObject classes it is accessible as self.STATES
```

In case additional states for a specific implementation are needed, there is a placeholder enumeration.

```
class DefaultSpecificState(enum.Enum):
    UNKNOWN = "UNKNOWN"
```
Similar to STATES, a SPECIFIC_STATES object allows to access the specific states, without knowing the name of the class where they are defined.

## VALUES Enum and how to expand it.
Another abstraction concept is the enumeration for the values, which can be get/set. It allows to define any value as an Enum, so only the name is fixed, but the value itself is flexible. It is defined in the AbstractNState.py with just one member:
```
class BaseValueEnum(enum.Enum):
    """Defines only the compulsory unknown."""

    UNKNOWN = "UNKNOWN"
```

> The VALUES object is asigned to BaseValueEnum. In the inheriting from AbstractNState classes it is overloaded by simply getting the information from the configuration file.

Reviewer 2: The VALUES object is as**s**igned to BaseValueEnum. In the inheriting from AbstractNState classes**,** it is **loaded** by simply getting the information from the configuration file **if it is present**.

For example, from the following configuration
```
<values>{"A10": (0, 10, 0.15), "A20": (1, 20, 0.3), "A30": (2, 30, 0.63), "A50": (3, 50, 0.9), "A75": (4, 75, 0.96)}</values>
```

The python code transforms the string from the configuration to an Enum:
```
values = ast.literal_eval(self.get_property("values"))
values_dict = dict(**{item.name: item.value for item in self.VALUES})
values_dict.update(values)
self.VALUES = Enum("ValueEnum", values_dict)
```
This is the same as if a ValueEnum class was defined as follows:
```
class ValueEnum(enum.enum):
   A10 = (0, 10, 0.15)
   A20 = (1, 20, 0.3)
   A30 = (2, 30, 0.63)
   A50 = (3, 50, 0.9)
   A75 = (4, 75, 0.96)
```

## Timeout
A timeout parameter can be added to any method, which needs it. The common behaviour is defined as:

 - timeout=0 - return immediately, without waiting
 - timeout=None - wait until the end of the action
 - timeout=n - wait for n seconds (float value). Raise an exception if error.

## Abstract Classes

> Only the defined or overloaded in the class methods and properties are listed. The exhaustive list of the available methods and properties for each class depeds on the parent class.

Reviewer 2: Typo: depends

Question: Is the word _overloaded_ means overwritten by values specified in a configuration?

> Only the defined or overloaded in the class methods and properties are listed. The exhaustive list of the available methods and properties for each class depeds on the parent class.

Reviewer 2: ```suggestion
Only the defined or overloaded in the class methods and properties are listed. The exhaustive list of the available methods and properties for each class depends on the parent class.
```

### AbstractActuator

The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractActuator.AbstractActuator` defines methods, common for any moving device.

Inherits from **HardwareObject** class.
  - defined methods:
    get_value, set_value, validate_value, update_value, _set_value, get_limits, update_limits, re_emit_values, force_emit_signals

>   - properties:

Reviewer: Do you mean 'properties' (as in 'defined in configuration files') or attributes, or what?

>     actuator_name, username, read_only, default_value, default_limits, unit

Reviewer 2: I am always very puzzled when I see **username**, can you explain because I see it in the xml configuration file.

Reviewer 2: @rhfogh @beteva This definitely with Rasmus configuration doc (which i enjoy) but i am lost at finding it. As a beamline scientist, some of the terminology you are all familiar with in the documentation goes over my head. I will try to come up to speed. I was heavily dependent on the examples provided either in the mockup directory or present in the modules. Rasmus, an example in you doc would help too.  If the ``device class`` and ``equipment class`` are obsolete, it would really push everyone in the right direction to have uptodated files in that directory, for example detector-distance-mockup.xml, or session.xml

Reviewer: @AnnieHeroux The text in my doc mentions two examples, to wit `beamline_config.yml` and `mxcubecore/configuration/mockup/detector-mockup.xml`. Would it help if I gave a full directory path also for beamline_config.yml? or do you think there shoudl be a direct link on the page? Or would we need more or better examples?

BTW, does anyone know if there is a way to link from a document in docs to a file in the mxcubecore repository?

Reviewer 3: > BTW, does anyone know if there is a way to link from a document in docs to a file in the mxcubecore repository?

@rhfogh, there should be. I will look for it again. Remind me if I forget to come back to you with an answer.

  - emited signals:
    valueChanged. stateChanged, limitsChanged

The _set_value is the only abstract method that needs to be overloaded with every specific implementation.

### AbstractEnergy

The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractEnergy.AbstractEnergy` handles the energy and wavelength.

Inherits from **AbstractActuator** class.
  - methods:
    get_wavelength, get_wavelength_limits, set_wavelength, calculate_energy, calculate_wavelength
  - properties:
    is_tunable
  - emited signals:
    valueChanged

### AbstractMotor

> The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractMotor.AbstractMotor` implement methods to handle devices which behave as a motor. Defines the MotorStaes enumeration, assigned to SPECIFIC_STATES.

Reviewer 2: MAybe: implements
Typo : MotorStates

> The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractMotor.AbstractMotor` implement methods to handle devices which behave as a motor. Defines the MotorStaes enumeration, assigned to SPECIFIC_STATES.

Reviewer 2: ```suggestion
The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractMotor.AbstractMotor` implements methods to handle devices which behave as a motor. Defines the MotorStates enumeration, assigned to SPECIFIC_STATES.
```

Inherits from **AbstractActuator** class.
  - methods:
    get_velocity, set_velocity, set_value_relative, home
  - properties:
    tolerance
  - emited signals:

>     valueChanged

Reviewer: The Superclass emits this, and also two more. Should we copy the signals in each class, or (maybe better) list them only in the superclass where they are defined, or what?

Author: Agree with you - we should only document whatever is different from the superclass, not repeat every time the methods, the signals...

Reviewer 2: Maybe I am not seeing here how everything is linked but when i want to have some documentation about a specific thing, you should not assume I already read everything else.

### AbstractNState

> The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractNState.AbstractNState` implemets methods to handle devices with a fixed number of possible values.

Reviewer 2: typo: implements

> The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractNState.AbstractNState` implemets methods to handle devices with a fixed number of possible values.

Reviewer 2: ```suggestion
The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractNState.AbstractNState` implements methods to handle devices with a fixed number of possible values.
```

Defines the BaseValueEnum, assigned to VALUES.

Inherits from **AbstractActuator** class.
  - methods:
    initialise_values, value_to_enum
  - overloaded methods:
    validate_value, set_limits, update_limits, re_emit_values

### AbstractFlux

> The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractFlux.AbstractFlux` handle the flux. Defines get_dose_rate_per_photon_per_mmsq as dose rate for a standard composition crystal, in Gy/s as a function of energy in keV.

Reviewer 2: typo: handles

> The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractFlux.AbstractFlux` handle the flux. Defines get_dose_rate_per_photon_per_mmsq as dose rate for a standard composition crystal, in Gy/s as a function of energy in keV.

Reviewer 2: ```suggestion
The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractFlux.AbstractFlux` handles the flux. Defines get_dose_rate_per_photon_per_mmsq as dose rate for a standard composition crystal, in Gy/s as a function of energy in keV.
```

There is some description in the file: would that be useful to add it in here??
https://github.com/mxcube/mxcubecore/blob/61da0cf4a7577d7b5af3bab2ecdbb6cfee26613a/mxcubecore/HardwareObjects/abstract/AbstractFlux.py#L46-L61

Inherits from **AbstractActuator** class.
  - methods:
    get_average_flux_density
  - properties:
    get_dose_rate_per_photon_per_mmsq

### AbstractTransmission

The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractTransmission.AbstractTransmission` get/set transmission value.

> Defines the MotorStaes enumeration, assigned to SPECIFIC_STATES.

Reviewer 2: Typo: MotorStates

> Defines the MotorStaes enumeration, assigned to SPECIFIC_STATES.

Reviewer 2: ```suggestion
Defines the MotorStates enumeration, assigned to SPECIFIC_STATES.
```

# Abstract Classes
## General Concept

The role of the mxcubecore Abstract Classes is to provide an API to be used by both mxcubeweb and mxcubeqt clients. It defines methods, attributes and signals. Most of the methods are defined only to give the footprint and should be, when appropriate, overloaded by the inheriting classes.

Abstract class can be inherited by any class, including another abstract class.

> Any abstract class, if not inheriting from another abstract class should inherit from the **HardwareObject** or alternatively **HardwareObjectYam** class. The [hierarchy scheme](https://github.com/mxcube/mxcubecore/blob/develop/Hierarchy.pdf) shows in brief the inheritance of the abstract classes.

Reviewer 4: typo: HardwareObjectYam -> HardwareObjectYaml

> If get_value/set_value methods are defined in the abstract class, the name of the class defines what will be represented. For example, get_value/set_value in the AbstractEnergy will read/set the energy, while the wavelength, handled in the same class, will have specific methods like get_value_wavelemgth/set_value_wavelength.

Reviewer 4: get_value_wavelemgth -> get_value_waveleNgth

## Methods

> There are two methods, which are available for any class, as inherited from HardwareObject - **get_object_by_role()** and **get_property()**. Both methods allow to retrieve information from the configuration file (xml or yaml)

Reviewer 4: maybe worth mentioning `get_properties`?

> - get_object_by_role returns an object, associated to a role. Used to give an access to the defined Hardware Object.

Reviewer 4: where is the role defined? Only in the config file?

  ```<object href="/aperture" role="beamsize"/>```
- get_property returns a property - numeric, string or None. This usually serves to define or asign different attributes.

  ```<values>{"IN": False, "OUT": True}</values>```

Methods, specific to different abstract classes will be listed further.

## STATES and SPECIFIC_STATES Enum
There is only a limited number of general states, shared by all HardwareObjects.
They are defined as an Enum as follows and should not be overridden.
```
@enum.unique
class HardwareObjectState(enum.Enum):
    """Enumeration of common states, shared between all HardwareObjects"""

    UNKNOWN = 0
    WARNING = 1
    BUSY = 2
    READY = 3
    FAULT = 4
    OFF = 5
```
The HardwareObjectState class is assigned to the STATES object, so it is accessible by the inheriting from HardwareObject classes as self.STATES.

> In case additional states for a specific implementation are needed, there is a placeholder enumeration.

Reviewer 4: any reason we should not override the `HardwareObjectState` if needed? (Instead of creating a new state object)

    UNKNOWN = "UNKNOWN"
```
The VALUES object is assigned to BaseValueEnum. In the inheriting from AbstractNState classes it is overloaded by simply getting the information from the configuration file.
For example, from the following configuration
```
<values>{"A10": (0, 10, 0.15), "A20": (1, 20, 0.3), "A30": (2, 30, 0.63), "A50": (3, 50, 0.9), "A75": (4, 75, 0.96)}</values>
```

The python code transforms the string from the configuration to an Enum:
```
values = ast.literal_eval(self.get_property("values"))
values_dict = dict(**{item.name: item.value for item in self.VALUES})
values_dict.update(values)
self.VALUES = Enum("ValueEnum", values_dict)
```
This is the same as if a ValueEnum class was defined as follows:
```
class ValueEnum(enum.enum):
   A10 = (0, 10, 0.15)
   A20 = (1, 20, 0.3)
   A30 = (2, 30, 0.63)
   A50 = (3, 50, 0.9)
   A75 = (4, 75, 0.96)
```

## Timeout
A timeout parameter can be added to any method, which needs it. The common behaviour is defined as:

 - timeout=0 - return immediately, without waiting
 - timeout=None - wait until the end of the action
 - timeout=n - wait for n seconds (float value). Raise an exception if error.

## Abstract Classes

> Only the defined or overloaded in the class methods, attributes, signals and properties are listed. The exhaustive list of the available methods and properties for each class depends on the parent class.

Reviewer 5: ```suggestion
Only the defined or overloaded methods, attributes, signals and properties are listed. The exhaustive list of the available methods and properties for each class depends on the parent class.
```

### AbstractActuator

The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractActuator.AbstractActuator` defines methods, common for any moving device.

Inherits from **HardwareObject** class.
  - defined methods:
    get_value, set_value, validate_value, update_value, _set_value, get_limits, update_limits
  - properties from the configuration file:
    actuator_name, username, read_only, default_value, default_limits
  - attributes:
    actuator_name, username, read_only, default_value, default_limits, unit, _nominal_value, _nominal_limits.
  - emitted signals:
    valueChanged, limitsChanged

The _set_value is the only abstract method that needs to be overloaded with every specific implementation.

### AbstractEnergy

The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractEnergy.AbstractEnergy` handles energy and wavelength.

Inherits from **AbstractActuator** class.
  - methods:
    get_wavelength, get_wavelength_limits, set_wavelength, calculate_energy, calculate_wavelength
  - overloaded methods:
    update_value, force_emit_signals
  - attributes:
    is_tunable
  - emitted signals:

>     energyChanged

Reviewer 4: add a deprecation note to this signal?

    definer, aperture, slits, _beam_divergence, _beam_position_on_screen, _beam_width, _beam_height, ._beam_shape, _beam_label, _beam_size_dict
  - emitted signals:
    beamSizeChanged. beamInfoChanged, beamPosChanged

### AbstractShutter

The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractShutter.AbstractShutter` implements methods to handle shutter like devices. Adds appropriate members to the BaseValueEnum.

Inherits from **AbstractNStater** class.
  - methods:
    open, close
  - attributes:
    is_open

## AbstractXRFSpectrum

The {py:class}`mxcubecore.HardwareObjects.abstract.AbstractXRFSpectrum.AbstractXRFSpectrum` implements methods to run XRF Spectrum type acquisition.

Inherits from **HardwareObject** class.
  - methods:
    start_xrf_spectrum, execute_xrf_spectrum, _execute_xrf_spectrum,
    spectrum_store_lims, spectrum_command_finished, spectrum_command_failed,
    spectrum_command_aborted, spectrum_status_change, spectrum_analyse,
    create_directory, get_filename.
  - properties from the configuration file:
    default_integration_time
  - attributes:
    lims, spectrum_info_dict, default_integration_time, spectrum_running
  - emitted signals:

>     xrfSpectrumStatusChanged

Reviewer 4: add a deprecation note to this signal?

> _execite_xrf_scan is the only abstract method. It is a placeholder for specific  sequence implementation.

Reviewer 4: execite_xrf_scan -> execute_xrf_scan
